contracts/DelegateContribute.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+interface IWVEOLAS {
+    /// @dev Gets the account balance in native token.
+    /// @param account Account address.
+    /// @return balance Account balance.
+    function balanceOf(address account) external view returns (uint256 balance);
+
+    /// @dev Gets the voting power.
+    /// @param account Account address.
+    /// @return balance Account balance.
+    function getVotes(address account) external view returns (uint256 balance);
+}
+
+/// @dev No balance available to delegate.
+/// @param delegator provided delegator address.
+error NoBalance(address delegator);
+
+/// @dev No self delegation.
+/// @param delegator provided delegator address.
+error NoSelfDelegation(address delegator);
+
+/// @title DynamicContribution - Ownable smart contract for minting ERC721 tokens
+contract DelegateContribute {
+
+    /// @dev Delegation.
+    /// @param delegator provided delegator.
+    /// @param delegatee provided delegatee.
+    event Delegation(address indexed delegator, address indexed delegatee);
+
+    // wveOLAS address
+    address immutable wveOLAS;
+
+    // Mapping of account address => address
+    mapping(address => address) public mapDelegation;
+
+    // Maps a delegatee to their list of delegators
+    mapping(address => address[]) private delegatorLists;
+
+    // Maps a delegator to their index in the delegatee's list + 1 (since default value is 0)
+    mapping(address => mapping(address => uint256)) private delegatorIndex;
+
+    /// @dev DynamicContribution constructor.
+    /// @param _wveOLAS wveOLAS contract address.
+    constructor(address _wveOLAS)
+    {
+        wveOLAS = _wveOLAS;
+    }
+
+    /// @dev Delegates Contribute voting power to an address.
+    /// @param delegatee Delegatee address.
+    function delegate(address delegatee) external {
+        address delegator = msg.sender;
+        if (delegator == delegatee) {
+            revert NoSelfDelegation(delegator);
+        }
+        uint256 balanceOf = IWVEOLAS(wveOLAS).balanceOf(delegator);
+        if (balanceOf == 0) {
+            revert NoBalance(delegator);
+        }
+
+        // Remove from old delegatee if applicable
+        address currentDelegatee = mapDelegation[delegator];
+        if(currentDelegatee != address(0) && currentDelegatee != delegatee) {
```

Comment:
I mean (*):
  address currentDelegatee = mapDelegation[delegator];
  if(currentDelegatee == delegatee) {
     revert NoDoubleDelegate(currentDelegatee, delegatee);
}

Reply:
It will be overwritten below, which doesn't make sense as well, so revert is better.

Reply:
Ah we are getting confused in word's meaning - `delegatee` is the one having multiple `delegator`-s.

Reply:
Here root of problem:
delegatorLists[delegatee].push(delegator); /// Array, not Set!

Try
call delegate(addr1);
call delegate(addr1);
As a result, you will add to the array address[] twice (n-times) msg.sender.
Same call delegate(0x0);
ref: mapping(address => address[]) private delegatorLists;

Reply:
you're right - fixed now